wgpu-core/src/instance.rs

```diff
@@ -262,25 +267,49 @@ impl<I: Clone> AdapterInputs<'_, I> {
     }
 }
 
+struct Adapters<F: IdentityFilter<AdapterId>> {
+    device_types: Vec<HalDeviceType>,
+    #[cfg(any(
+        not(any(target_os = "ios", target_os = "macos")),
+        feature = "gfx-backend-vulkan"
+    ))]
+    vulkan: (
+        Vec<HalAdapter<gfx_backend_vulkan::Backend>>,
+        Option<F::Input>,
+    ),
+    #[cfg(any(target_os = "ios", target_os = "macos"))]
+    metal: (
+        Vec<HalAdapter<gfx_backend_metal::Backend>>,
+        Option<F::Input>,
+    ),
+    #[cfg(windows)]
+    dx12: (Vec<HalAdapter<gfx_backend_dx12::Backend>>, Option<F::Input>),
```

Comment:
Instead of returning a tuple here and accessing with `.0` and `.1`, I could make small structure like this:
```rust
struct BackendData<F: IdentityFilter<AdapterId>> {
    adapters: Vec<HalAdapter<gfx_backend_dx12::Backend>>,
    id: Option<F::Input>,
}
```
If desired.